internal/custom_image/custom_image_data_source.go

```diff
@@ -0,0 +1,272 @@
+package custom_image
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+
+	swagger "github.com/crusoecloud/client-go/swagger/v1alpha5"
+	"github.com/crusoecloud/terraform-provider-crusoe/internal/common"
+)
+
+type customImageDataSource struct {
+	client     *swagger.APIClient
+	basePath   string
+	httpClient *http.Client
+}
+
+type customImageDataSourceModel struct {
+	Name         types.String       `tfsdk:"name"`
+	NamePrefix   types.String       `tfsdk:"name_prefix"`
+	CustomImages []customImageModel `tfsdk:"custom_images"`
+	NewestImage  *customImageModel  `tfsdk:"newest_image"`
+}
+
+type customImageModel struct {
+	ID          string `tfsdk:"id" json:"id"`
+	Name        string `tfsdk:"name" json:"name"`
+	Description string `tfsdk:"description" json:"description"`
+	Location    string `tfsdk:"location" json:"location"`
+	Status      string `tfsdk:"status" json:"status"`
+	CreatedAt   string `tfsdk:"created_at" json:"created_at"`
+	UpdatedAt   string `tfsdk:"updated_at" json:"updated_at"`
+}
```

Comment:
This type doesn't match the data that the API will return:
- `Status` & `UpdatedAt` fields should be removed
- `Location` can probably stay, but the API technically supports a one:many association between custom images and locations. Use `swagger.Image.Locations[0]`, or modify the model to support `Locations []string`
- `Tags []string` may be used in place of the `Status` field